### ISSUE TITLE
Output for tfrecords files is not reproducible

### ISSUE THREAD

**Issue by leachim** [tensorflow/tensorflow#3640]
### Environment info
Operating System:
OS: Debian testing stretch                                
Kernel: x86_64 Linux 4.5.0-1-amd64

Tensorflow
Version: 0.9.0

### Steps to reproduce
Running this script gives a ~/test.tfrecords file. 
Running the script repeatedly gives different hashes for the file. 

```python
import os
import argparse

import tensorflow as tf
import numpy as np

MAXIMUM_SEQUENCE_LENGTH = 1000

def export(data, name, args):
    num_examples = len(data)

    fh_output = os.path.join(os.path.expandvars(args.output_directory), name + '.tfrecords')
    print('Writing', fh_output)
    writer = tf.python_io.TFRecordWriter(fh_output)
    for index in range(num_examples):
        ## prepare things to save
        seq_x1 = data[index][0].tostring()
        seq_x2 = data[index][1].tostring()

        example = tf.train.Example(features=tf.train.Features(feature={
            'x1': _bytes_feature(seq_x1),
            'x2': _bytes_feature(seq_x2),
            }))
        writer.write(example.SerializeToString())
    writer.close()

def _int64_feature(value):
    return tf.train.Feature(int64_list=tf.train.Int64List(value=[value]))

def _bytes_feature(value):
    return tf.train.Feature(bytes_list=tf.train.BytesList(value=[value]))


if __name__ == '__main__':
    tf.set_random_seed(11)
    np.random.seed(11)

    parser = argparse.ArgumentParser()
    parser.add_argument('--output_directory', type=str, default='$HOME/', help='Directory to write the converted result')
    # parser.add_argument('--validation_size', type=int, default=3000, help='Number of examples to separate from the training data for the validation set.')
    args = parser.parse_args()

    train = [( np.zeros(1000), np.ones(1000), 500 )]

    ## Write data to training file
    export(train, 'test', args)
```

I wonder why I cannot produce the same binary file again, if I rerun the script. Does this mean the content is different? I didn't find any option to set a seed either.

**Comment by poxvoculi** [comment]
When I run this program I get the same result every time, with identical md5sum hashes.  What happens if you print the protobuf, instead of using SerializeToString(), are they different?

**Comment by leachim** [comment]
I am very sorry, in my haste, I seem to have given you a bad toy example. Very strangely, in my setup if I execute a slightly modified example from above, I get different hashes: 
`
example = tf.train.Example(features=tf.train.Features(feature={
'x1': _bytes_feature(seq_x1),
'x2': _bytes_feature(seq_x2),
'x3': _bytes_feature(seq_x2),
}))
`
It seems that the protobuf is identical though, I have attached an example of the printed output of example (with x1,x2,x3), for which I obtain identical hashes for the protobuf in all runs. 
`print(example) > test.txt`
[test.txt](https://github.com/tensorflow/tensorflow/files/402820/test.txt)

md5sum test.tfrecords test.txt
1630f4da9a52856ac31e84fd482aebe0  test.tfrecords
9e9cda7528deb2e7aa358b39c56b2f1d  test.txt

md5sum test.tfrecords test.txt
e0af3c621d636f147e91582a00cc0834  test.tfrecords
9e9cda7528deb2e7aa358b39c56b2f1d  test.txt

The hashes for the tfrecords file start to repeat themselves as well. I am not entirely sure, whether a similar pattern happens for the x1, x2 case, with lower frequency, though. 

Can this be a problem in the following line then? 
`writer.write(example.SerializeToString())`

Can you reproduce the issue with the updated example? 
I could also try with a more up-to-date version of tensorflow alternatively.

**Comment by poxvoculi** [comment]
Sorry for the latency, just got a chance to run your updated program, with the three examples.  I ran it 6 times and got identical output files each time.  

You say you always see the same values printed by
  print(example)
Have you verified that by taking a checksum of that printed output?

If that is true, yet the example.SerializeToString() output varies, I would suspect a problem not with TensorFlow, but with the python implementation of protocol buffers that you're using, or maybe the filesystem.  It would be good to track down exactly which bytes are varying, and whether they're semantically significant (i.e. do they effect the reparsed value).

**Comment by leachim** [comment]
Ok. I managed to reproduce the issue on a digital ocean droplet. I have attached all the files, hope that helps to clarify things. 

System info: 
4.4.0-31-generic #50-Ubuntu SMP Wed Jul 13 00:07:12 UTC 2016
Ubuntu 16.04 xenia

Steps to reproduce: 
see command_info and issue.txt

I have also included the tfrecord files in the zip folder. 

[command_info.txt](https://github.com/tensorflow/tensorflow/files/404971/command_info.txt)
[issue.txt](https://github.com/tensorflow/tensorflow/files/404974/issue.txt)
[issue.zip](https://github.com/tensorflow/tensorflow/files/404975/issue.zip)

**Comment by poxvoculi** [comment]
Doh!  Sorry for not seeing the issue here sooner.  tf.train.example is a protocol message whose internal representation uses some kind of map.  The order in which the map entries are stored in serializied-format is non-deterministic, so that's where your checksum differences come from, even though the ASCII serializations are identical.  This should be clear enough if you squint at your example1.tfrecords and example2.tfrecords: You can see three records inside which look approximately the same, but are in different order.

The non-deterministic ordering of maps inside protobufs is a feature of the protobuf implementation, not TensorFlow.

**Comment by jhagege** [comment]
Did you find a solution for reproducibility of TFRecords creation, especially when it is multithreaded ? 
Thanks !
